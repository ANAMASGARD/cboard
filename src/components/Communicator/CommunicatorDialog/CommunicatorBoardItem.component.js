import React from 'react';
import PropTypes from 'prop-types';
import { intlShape } from 'react-intl';
import MenuIcon from '@material-ui/icons/MoreVert';
import AddIcon from '@material-ui/icons/AddCircle';
import RemoveIcon from '@material-ui/icons/RemoveCircle';
import PublicIcon from '@material-ui/icons/Public';
import KeyIcon from '@material-ui/icons/VpnKey';
import BrokenIcon from '@material-ui/icons/BrokenImage';
import IconButton from '../../UI/IconButton';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import { TAB_INDEXES } from './CommunicatorDialog.constants';
import messages from './CommunicatorDialog.messages';

class CommunicatorBoardItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      menu: null
    };
  }

  openMenu(e) {
    this.setState({ menu: e.currentTarget });
  }

  closeMenu() {
    this.setState({ menu: null });
  }

  render() {
    const { board, selectedTab, intl } = this.props;
    const title = intl.formatMessage({
      id: board.nameKey || board.name || board.id
    });

    return (
      <div className="CommunicatorDialog__boards__item">
        <div className="CommunicatorDialog__boards__item__image">
          {!!board.caption && <img src={board.caption} alt={title} />}
          {!board.caption && (
            <div className="CommunicatorDialog__boards__item__image__empty">
              <BrokenIcon />
            </div>
          )}
          <div className="CommunicatorDialog__boards__item__image__button">
            <IconButton
              label={intl.formatMessage(
                selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS
                  ? messages.removeBoard
                  : messages.addBoard
              )}
              onClick={() => {}}
            >
              {selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS ? (
                <RemoveIcon />
              ) : (
                <AddIcon />
              )}
            </IconButton>
          </div>
        </div>
        <div className="CommunicatorDialog__boards__item__data">
          <div className="CommunicatorDialog__boards__item__data__button">
            <IconButton
              className={`board-item-menu-button board-item-menu-button-${
                board.id
              }`}
              label={intl.formatMessage(messages.menu)}
              onClick={this.openMenu.bind(this)}
            >
              <MenuIcon />
            </IconButton>
            <Menu
              className={`board-item-menu board-item-menu-${board.id}`}
              anchorEl={this.state.menu}
              open={Boolean(this.state.menu)}
              onClose={this.closeMenu.bind(this)}
            >
              <MenuItem onClick={() => {}}>...</MenuItem>
            </Menu>
          </div>

          <div className="CommunicatorDialog__boards__item__data__title">
            {title}
          </div>
          <div className="CommunicatorDialog__boards__item__data__author">
            {intl.formatMessage(messages.author, { author: board.author })}
          </div>
          <div className="CommunicatorDialog__boards__item__data__extra">
            {selectedTab === TAB_INDEXES.ALL_BOARDS && <PublicIcon />}
            {selectedTab === TAB_INDEXES.MY_BOARDS &&
              board.isPublic && <PublicIcon />}
            {selectedTab === TAB_INDEXES.MY_BOARDS &&
              !board.isPublic && <KeyIcon />}
          </div>
        </div>
      </div>
    );
  }
}

CommunicatorBoardItem.propTypes = {
  intl: intlShape,
  selectedTab: PropTypes.number,
  board: PropTypes.object
};

export default CommunicatorBoardItem;
