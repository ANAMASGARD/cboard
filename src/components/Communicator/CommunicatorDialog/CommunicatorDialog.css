.CommunicatorDialog__container {
  width: 100%;
  height: calc(100vh - 70px);
  min-height: 300px;
  overflow: hidden;
  padding: 0;
}

.CommunicatorDialog__tabs {
  background-color: #3f51b5;
  color: #fff;
}

.CommunicatorDialog__tabs [class^='TabIndicator-'] {
  background-color: #fff;
  height: 3px;
}

.CommunicatorDialog__content {
  height: calc(100% - 48px);
  overflow: auto;
  padding: 0;
}

.CommunicatorDialog__content .CommunicatorDialog__spinner {
  margin: 40px auto;
  display: block;
}

.CommunicatorDialog__communicatorData {
  background: #eee;
  padding: 16px 16px 28px;
}

.CommunicatorDialog__communicatorData__title {
  font-size: 16px;
}

.CommunicatorDialog__communicatorData__boardsQty {
  font-size: 12px;
  margin-top: 5px;
  color: #888;
}

.CommunicatorDialog__boards {
  background: #fff;
  padding: 20px 16px;
}

.CommunicatorDialog__boards__item {
  display: flex;
  min-height: 150px;
  margin-bottom: 10px;
}

.CommunicatorDialog__boards__item__data {
  position: relative;
  width: 60%;
  padding-left: 20px;
}

.CommunicatorDialog__boards__item__image {
  position: relative;
  width: 40%;
}

.CommunicatorDialog__boards__item__image__empty {
  height: 100%;
  background: #eee;
  position: relative;
}

.CommunicatorDialog__boards__item__image__empty svg {
  position: absolute;
  font-size: 48px;
  color: #aaa;
  top: calc(50% - 24px);
  left: calc(50% - 24px);
}
.CommunicatorDialog__boards__item__image__button {
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: 10;
}

.CommunicatorDialog__boards__item__image__button svg {
  font-size: 40px;
}

.CommunicatorDialog__boards__item__data__button {
  position: absolute;
  top: 0;
  right: 0;
  z-index: 10;
}

.CommunicatorDialog__boards__item__data__title,
.CommunicatorDialog__boards__item__data__author,
.CommunicatorDialog__boards__item__data__extra {
  margin-top: 5px;
}
