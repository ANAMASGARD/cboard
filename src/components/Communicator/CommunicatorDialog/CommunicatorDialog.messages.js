import { defineMessages } from 'react-intl';

export default defineMessages({
  title: {
    id: 'cboard.components.CommunicatorDialog.title',
    defaultMessage: 'My Communicator'
  },
  menu: {
    id: 'cboard.components.CommunicatorDialog.menu',
    defaultMessage: 'Menu'
  },
  menuRootBoardOption: {
    id: 'cboard.components.CommunicatorDialog.menuRootBoardOption',
    defaultMessage: 'Set as Root Board'
  },
  menuPublishOption: {
    id: 'cboard.components.CommunicatorDialog.menuPublishOption',
    defaultMessage: 'Publish Board'
  },
  menuUnpublishOption: {
    id: 'cboard.components.CommunicatorDialog.menuUnpublishOption',
    defaultMessage: 'Unpublish Board'
  },
  emptyBoardsList: {
    id: 'cboard.components.CommunicatorDialog.emptyBoardsList',
    defaultMessage: 'No boards :('
  },
  loadNextPage: {
    id: 'cboard.components.CommunicatorDialog.loadNextPage',
    defaultMessage: 'Load More'
  },
  search: {
    id: 'cboard.components.CommunicatorDialog.search',
    defaultMessage: 'Search'
  },
  author: {
    id: 'cboard.components.CommunicatorDialog.author',
    defaultMessage: 'By {author}'
  },
  addBoard: {
    id: 'cboard.components.CommunicatorDialog.addBoard',
    defaultMessage: 'Add Board'
  },
  removeBoard: {
    id: 'cboard.components.CommunicatorDialog.removeBoard',
    defaultMessage: 'Remove Board'
  },
  communicatorBoards: {
    id: 'cboard.components.CommunicatorDialog.communicatorBoards',
    defaultMessage: 'Boards'
  },
  allBoards: {
    id: 'cboard.components.CommunicatorDialog.allBoards',
    defaultMessage: 'Public Boards'
  },
  myBoards: {
    id: 'cboard.components.CommunicatorDialog.myBoards',
    defaultMessage: 'All My Boards'
  },
  boardsQty: {
    id: 'cboard.components.CommunicatorDialog.boardsQty',
    defaultMessage: '{qty} boards'
  },
  helpAndSupport: {
    id: 'cboard.components.CommunicatorDialog.helpAndSupport',
    defaultMessage: 'Help And Support'
  },
  termsOfService: {
    id: 'cboard.components.CommunicatorDialog.termsOfService',
    defaultMessage: 'Terms Of Service'
  }
});
