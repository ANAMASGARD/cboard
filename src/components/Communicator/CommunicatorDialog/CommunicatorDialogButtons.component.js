import React from 'react';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import SearchIcon from '@material-ui/icons/Search';
import IconButton from '../../UI/IconButton';
import MenuIcon from '@material-ui/icons/MoreVert';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import { intlShape } from 'react-intl';

import messages from './CommunicatorDialog.messages';

class CommunicatorDialogButtons extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      searchValue: '',
      showSearchBar: false,
      menu: null
    };
  }

  openMenu(e) {
    this.setState({ menu: e.currentTarget });
  }

  closeMenu() {
    this.setState({ menu: null });
  }

  openSearchBar() {
    this.setState({ showSearchBar: true });
  }

  onSearch(event) {
    const searchValue = event.target.value;

    this.setState({ searchValue });
    this.props.onSearch(searchValue);
  }

  render() {
    const { intl } = this.props;

    return (
      <div className="CommunicatorDialogButtons__container">
        {this.state.showSearchBar && (
          <div className="CommunicatorDialogButtons__searchInput">
            <TextField
              id="communicator-search"
              value={this.state.searchValue}
              onChange={this.onSearch.bind(this)}
              margin="normal"
            />
          </div>
        )}
        {!this.state.showSearchBar && (
          <div className="CommunicatorDialogButtons__searchButton">
            <IconButton
              label={intl.formatMessage(messages.search)}
              onClick={this.openSearchBar.bind(this)}
            >
              <SearchIcon />
            </IconButton>
          </div>
        )}

        <div className="CommunicatorDialogButtons__menu">
          <IconButton
            label={intl.formatMessage(messages.menu)}
            onClick={this.openMenu.bind(this)}
          >
            <MenuIcon />
          </IconButton>
          <Menu
            id="communicator-dialog-buttons-menu-button"
            anchorEl={this.state.menu}
            open={Boolean(this.state.menu)}
            onClose={this.closeMenu.bind(this)}
          >
            <MenuItem onClick={() => {}}>...</MenuItem>
          </Menu>
        </div>
      </div>
    );
  }
}

CommunicatorDialogButtons.propTypes = {
  onSearch: PropTypes.func.isRequired,
  intl: intlShape.isRequired
};

export default CommunicatorDialogButtons;
