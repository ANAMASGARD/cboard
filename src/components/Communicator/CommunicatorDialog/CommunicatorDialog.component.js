import React from 'react';
import PropTypes from 'prop-types';
import FullScreenDialog, {
  FullScreenDialogContent
} from '../../UI/FullScreenDialog';
import Paper from '@material-ui/core/Paper';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import CircularProgress from '@material-ui/core/CircularProgress';
import MenuIcon from '@material-ui/icons/MoreVert';
import AddIcon from '@material-ui/icons/AddCircle';
import RemoveIcon from '@material-ui/icons/RemoveCircle';
import PublicIcon from '@material-ui/icons/Public';
import KeyIcon from '@material-ui/icons/VpnKey';
import BrokenIcon from '@material-ui/icons/BrokenImage';
import IconButton from '../../UI/IconButton';
import { intlShape } from 'react-intl';

import { TAB_INDEXES } from './CommunicatorDialog.constants';
import messages from './CommunicatorDialog.messages';

import './CommunicatorDialog.css';

const CommunicatorDialog = ({
  open,
  intl,
  selectedTab,
  loading,
  boards,
  onClose,
  onTabChange
}) => (
  <FullScreenDialog
    disableSubmit={true}
    open={open}
    title={intl.formatMessage(messages.title)}
    onClose={onClose}
  >
    <Paper>
      <FullScreenDialogContent className="CommunicatorDialog__container">
        <Tabs
          value={selectedTab}
          onChange={onTabChange}
          className="CommunicatorDialog__tabs"
          fixed
          scrollable
          scrollButtons="off"
        >
          <Tab
            label={intl.formatMessage(messages.communicatorBoards)}
            className={
              selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS ? 'active' : ''
            }
          />
          <Tab
            label={intl.formatMessage(messages.allBoards)}
            className={selectedTab === TAB_INDEXES.ALL_BOARDS ? 'active' : ''}
          />
          <Tab
            label={intl.formatMessage(messages.myBoards)}
            className={selectedTab === TAB_INDEXES.MY_BOARDS ? 'active' : ''}
          />
        </Tabs>

        <div className="CommunicatorDialog__content">
          {!loading && (
            <React.Fragment>
              {selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS && (
                <div className="CommunicatorDialog__communicatorData">
                  <React.Fragment>
                    <div className="CommunicatorDialog__communicatorData__title">
                      {intl.formatMessage(messages.title)}
                    </div>
                    <div className="CommunicatorDialog__communicatorData__boardsQty">
                      {intl.formatMessage(messages.boardsQty, {
                        qty: boards.length
                      })}
                    </div>
                  </React.Fragment>
                </div>
              )}

              <div className="CommunicatorDialog__boards">
                {boards.map((board, i) => {
                  const title = intl.formatMessage({
                    id: board.nameKey || board.name || board.id
                  });
                  return (
                    <div key={i} className="CommunicatorDialog__boards__item">
                      <div className="CommunicatorDialog__boards__item__image">
                        {!!board.caption && (
                          <img src={board.caption} alt={title} />
                        )}
                        {!board.caption && (
                          <div className="CommunicatorDialog__boards__item__image__empty">
                            <BrokenIcon />
                          </div>
                        )}
                        <div className="CommunicatorDialog__boards__item__image__button">
                          <IconButton
                            label={intl.formatMessage(
                              selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS
                                ? messages.removeBoard
                                : messages.addBoard
                            )}
                            onClick={() => {}}
                          >
                            {selectedTab === TAB_INDEXES.COMMUNICATOR_BOARDS ? (
                              <RemoveIcon />
                            ) : (
                              <AddIcon />
                            )}
                          </IconButton>
                        </div>
                      </div>
                      <div className="CommunicatorDialog__boards__item__data">
                        <div className="CommunicatorDialog__boards__item__data__button">
                          <IconButton
                            label={intl.formatMessage(messages.menu)}
                            onClick={() => {}}
                          >
                            <MenuIcon />
                          </IconButton>
                        </div>

                        <div className="CommunicatorDialog__boards__item__data__title">
                          {title}
                        </div>
                        <div className="CommunicatorDialog__boards__item__data__author">
                          {board.author}
                        </div>
                        <div className="CommunicatorDialog__boards__item__data__extra">
                          {selectedTab === TAB_INDEXES.ALL_BOARDS && (
                            <PublicIcon />
                          )}
                          {selectedTab === TAB_INDEXES.MY_BOARDS &&
                            board.isPublic && <PublicIcon />}
                          {selectedTab === TAB_INDEXES.MY_BOARDS &&
                            !board.isPublic && <KeyIcon />}
                        </div>
                      </div>
                    </div>
                  );
                })}
              </div>
            </React.Fragment>
          )}

          {loading && (
            <CircularProgress
              size={25}
              className="CommunicatorDialog__spinner"
              thickness={7}
            />
          )}
        </div>
      </FullScreenDialogContent>
    </Paper>
  </FullScreenDialog>
);

CommunicatorDialog.defaultProps = {
  open: false,
  loading: false,
  selectedTab: 0,
  boards: [],
  onClose: () => {},
  onTabChange: () => {}
};

CommunicatorDialog.propTypes = {
  boards: PropTypes.array,
  open: PropTypes.bool,
  loading: PropTypes.bool,
  selectedTab: PropTypes.number,
  intl: intlShape,
  onClose: PropTypes.func,
  onTabChange: PropTypes.func
};

export default CommunicatorDialog;
