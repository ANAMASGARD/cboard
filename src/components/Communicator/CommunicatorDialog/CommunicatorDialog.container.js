import React from 'react';
import { connect } from 'react-redux';
import CommunicatorDialog from './CommunicatorDialog.component';
import { TAB_INDEXES } from './CommunicatorDialog.constants';
import { injectIntl } from 'react-intl';
import API from '../../../api';

const BOARDS_PAGE_LIMIT = 10;
const INITIAL_STATE = {
  publicBoards: {
    page: 0,
    total: 0,
    search: '',
    data: []
  },
  myBoards: {
    page: 0,
    total: 0,
    search: '',
    data: []
  }
};

const findLocalBoards = (boards, intl, value = '') => {
  return boards.filter(board => {
    const title = intl.formatMessage({
      id: board.nameKey || board.name || board.id
    });

    let returnValue = title.toLowerCase().indexOf(value.toLowerCase()) >= 0;
    returnValue =
      returnValue ||
      (board.author &&
        board.author.toLowerCase().indexOf(value.toLowerCase()) >= 0);

    return returnValue;
  });
};

const STATE_TAB_MAP = {
  [TAB_INDEXES.COMMUNICATOR_BOARDS]: 'communicatorBoards',
  [TAB_INDEXES.ALL_BOARDS]: 'publicBoards',
  [TAB_INDEXES.MY_BOARDS]: 'myBoards'
};

class CommunicatorDialogContainer extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
      boards: props.communicatorBoards, // First time => Communicator Boards Tab
      selectedTab: TAB_INDEXES.COMMUNICATOR_BOARDS,
      communicatorBoards: props.communicatorBoards,
      cboardBoards: props.cboardBoards,
      publicBoards: INITIAL_STATE.publicBoards,
      myBoards: INITIAL_STATE.myBoards,
      totalPages: Math.ceil(
        props.communicatorBoards.length / BOARDS_PAGE_LIMIT
      ),
      page: 1,
      search: ''
    };
  }

  async onTabChange(event, selectedTab = TAB_INDEXES.COMMUNICATOR_BOARDS) {
    const tabData = await this.doSearch(this.state.search, 1, 0, selectedTab);
    this.setState({
      ...tabData,
      selectedTab,
      page: 1
    });
  }

  async loadNextPage() {
    const page = this.state.page + 1;
    const selectedTabData = this.state[STATE_TAB_MAP[this.state.selectedTab]];
    const nextApiPage = selectedTabData.page + 1;
    const apiPages = Math.ceil(selectedTabData.total / BOARDS_PAGE_LIMIT);

    let newState = { page };
    let localPages = 0;
    if (this.state.selectedTab === TAB_INDEXES.ALL_BOARDS) {
      const localBoards = findLocalBoards(
        this.state.cboardBoards,
        this.props.intl,
        this.state.search
      );
      localPages = Math.ceil(localBoards.length / BOARDS_PAGE_LIMIT);
    }

    if (page > localPages && nextApiPage <= apiPages) {
      const {
        boards,
        totalPages,
        publicBoards,
        myBoards
      } = await this.doSearch(selectedTabData.search, nextApiPage);

      newState = {
        ...newState,
        boards,
        totalPages,
        publicBoards,
        myBoards
      };
    }

    this.setState(newState);
  }

  async doSearch(
    search,
    page = 1,
    offset = 0,
    selectedTab = this.state.selectedTab
  ) {
    let boards = [];
    let totalPages = 1;
    const selectedProperty = STATE_TAB_MAP[selectedTab];
    let dataForProperty =
      page > 1 ? this.state[selectedProperty] : INITIAL_STATE[selectedProperty];

    switch (selectedTab) {
      case TAB_INDEXES.COMMUNICATOR_BOARDS:
        dataForProperty = this.state.communicatorBoards;
        boards = findLocalBoards(dataForProperty, this.props.intl, search);
        totalPages = Math.ceil(boards.length / BOARDS_PAGE_LIMIT);
        break;

      case TAB_INDEXES.ALL_BOARDS:
        const localBoards = findLocalBoards(
          this.state.cboardBoards,
          this.props.intl,
          search
        );
        const publicBoardsResponse = await API.getBoards({
          limit: BOARDS_PAGE_LIMIT,
          page,
          search,
          offset
        });
        const totalAllBoards = localBoards.length + publicBoardsResponse.total;
        totalPages = Math.ceil(totalAllBoards / BOARDS_PAGE_LIMIT);
        dataForProperty = {
          ...publicBoardsResponse,
          data: dataForProperty.data.concat(publicBoardsResponse.data)
        };
        boards = localBoards.concat(dataForProperty.data);
        break;

      case TAB_INDEXES.MY_BOARDS:
        const myBoardsResponse = await API.getMyBoards({
          limit: BOARDS_PAGE_LIMIT,
          page,
          search,
          offset
        });
        totalPages = Math.ceil(myBoardsResponse.total / BOARDS_PAGE_LIMIT);
        dataForProperty = {
          ...myBoardsResponse,
          data: dataForProperty.data.concat(myBoardsResponse.data)
        };
        boards = dataForProperty.data;
        break;

      default:
        break;
    }

    const myBoards = this.state.myBoards;
    const publicBoards = this.state.publicBoards;
    return {
      boards,
      totalPages,
      publicBoards,
      myBoards,
      [selectedProperty]: dataForProperty
    };
  }

  async onSearch(search = this.state.search) {
    this.setState({
      boards: [],
      loading: true,
      page: 1,
      totalPages: 1,
      search
    });
    const { boards, totalPages, publicBoards, myBoards } = await this.doSearch(
      search
    );
    this.setState({
      boards,
      page: 1,
      totalPages,
      publicBoards,
      myBoards,
      loading: false
    });
  }

  render() {
    const limit = this.state.page * BOARDS_PAGE_LIMIT;
    const dialogProps = {
      ...this.props,
      ...this.state,
      limit,
      loadNextPage: this.loadNextPage.bind(this),
      onTabChange: this.onTabChange.bind(this),
      onSearch: this.onSearch.bind(this)
    };

    return <CommunicatorDialog {...dialogProps} />;
  }
}

const mapStateToProps = ({ board, communicator, language, app }, ownProps) => {
  const activeCommunicatorId = communicator.activeCommunicatorId;
  const currentCommunicator = communicator.communicators.find(
    communicator => communicator.id === activeCommunicatorId
  );

  const communicatorBoards = board.boards.filter(
    board => currentCommunicator.boards.indexOf(board.id) >= 0
  );

  const { userData } = app;
  const cboardBoards = board.boards;

  return {
    ...ownProps,
    communicatorBoards,
    cboardBoards,
    userData
  };
};

const mapDispatchToProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectIntl(CommunicatorDialogContainer));
