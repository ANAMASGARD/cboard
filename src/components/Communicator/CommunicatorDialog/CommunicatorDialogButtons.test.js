import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import CommunicatorDialogButtons from './CommunicatorDialogButtons.component';
import { TAB_INDEXES } from './CommunicatorDialog.constants';

jest.mock('./CommunicatorDialog.messages', () => {
  return {
    title: {
      id: 'cboard.components.CommunicatorDialog.title',
      defaultMessage: 'My Communicator'
    },
    menu: {
      id: 'cboard.components.CommunicatorDialog.menu',
      defaultMessage: 'Menu'
    },
    search: {
      id: 'cboard.components.CommunicatorDialog.search',
      defaultMessage: 'Search'
    },
    author: {
      id: 'cboard.components.CommunicatorDialog.author',
      defaultMessage: 'By {author}'
    },
    addBoard: {
      id: 'cboard.components.CommunicatorDialog.addBoard',
      defaultMessage: 'Add Board'
    },
    removeBoard: {
      id: 'cboard.components.CommunicatorDialog.removeBoard',
      defaultMessage: 'Remove Board'
    },
    communicatorBoards: {
      id: 'cboard.components.CommunicatorDialog.communicatorBoards',
      defaultMessage: 'Included Boards'
    },
    allBoards: {
      id: 'cboard.components.CommunicatorDialog.allBoards',
      defaultMessage: 'All Boards'
    },
    myBoards: {
      id: 'cboard.components.CommunicatorDialog.myBoards',
      defaultMessage: 'My Boards'
    },
    boardsQty: {
      id: 'cboard.components.CommunicatorDialog.boardsQty',
      defaultMessage: '{qty} boards'
    }
  };
});

const intlMock = {
  formatMessage: ({ id }) => id
};

let searchValue = '';
const COMPONENT_PROPS = {
  intl: intlMock,
  onSearch: value => {
    searchValue = value;
  }
};

describe('CommunicatorDialogButtons tests', () => {
  beforeEach(() => {
    searchValue = '';
  });

  test('default renderer', () => {
    shallowMatchSnapshot(<CommunicatorDialogButtons {...COMPONENT_PROPS} />);
  });

  test('search behavior', () => {
    let wrapper = shallow(<CommunicatorDialogButtons {...COMPONENT_PROPS} />);
    let tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    let currentState = wrapper.state();
    expect(currentState.showSearchBar).toBe(false);
    expect(currentState.searchValue).toBe('');

    const searchButton = wrapper.find(
      '#communicator-dialog-buttons-search-button'
    );
    searchButton.simulate('click');

    currentState = wrapper.state();
    expect(currentState.showSearchBar).toBe(true);
    expect(searchValue).toBe('');

    const searchInput = wrapper.find('#communicator-dialog-buttons-search');
    searchInput.simulate('change', { target: { value: 'something' } });

    currentState = wrapper.state();
    expect(currentState.showSearchBar).toBe(true);
    expect(searchValue).toBe('something');
  });

  test('menu behavior', () => {
    const wrapper = shallow(<CommunicatorDialogButtons {...COMPONENT_PROPS} />);
    let tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    let menu = wrapper.find('#communicator-dialog-buttons-menu').get(0);
    expect(menu.props.open).toBe(false);

    const menuButton = wrapper.find('#communicator-dialog-buttons-menu-button');
    menuButton.simulate('click', { currentTarget: 'someElement' });

    menu = wrapper.find('#communicator-dialog-buttons-menu').get(0);
    expect(menu.props.open).toBe(true);

    tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
