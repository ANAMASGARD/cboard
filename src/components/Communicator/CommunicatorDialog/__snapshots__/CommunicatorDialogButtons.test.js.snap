// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunicatorDialogButtons tests default renderer 1`] = `
<div
  className="CommunicatorDialogButtons__container"
>
  <div
    className="CommunicatorDialogButtons__searchButton"
  >
    <IconButton
      id="communicator-dialog-buttons-search-button"
      label="cboard.components.CommunicatorDialog.search"
      onClick={[Function]}
    >
      <pure(Search) />
    </IconButton>
  </div>
  <div
    className="CommunicatorDialogButtons__menu"
  >
    <IconButton
      id="communicator-dialog-buttons-menu-button"
      label="cboard.components.CommunicatorDialog.menu"
      onClick={[Function]}
    >
      <pure(MoreVert) />
    </IconButton>
    <WithStyles(Menu)
      anchorEl={null}
      id="communicator-dialog-buttons-menu"
      onClose={[Function]}
      open={false}
    >
      <WithStyles(MenuItem)
        onClick={[Function]}
      >
        ...
      </WithStyles(MenuItem)>
    </WithStyles(Menu)>
  </div>
</div>
`;

exports[`CommunicatorDialogButtons tests menu behavior 1`] = `
<div
  className="CommunicatorDialogButtons__container"
>
  <div
    className="CommunicatorDialogButtons__searchButton"
  >
    <IconButton
      id="communicator-dialog-buttons-search-button"
      label="cboard.components.CommunicatorDialog.search"
      onClick={[Function]}
    >
      <pure(Search) />
    </IconButton>
  </div>
  <div
    className="CommunicatorDialogButtons__menu"
  >
    <IconButton
      id="communicator-dialog-buttons-menu-button"
      label="cboard.components.CommunicatorDialog.menu"
      onClick={[Function]}
    >
      <pure(MoreVert) />
    </IconButton>
    <WithStyles(Menu)
      anchorEl={null}
      id="communicator-dialog-buttons-menu"
      onClose={[Function]}
      open={false}
    >
      <WithStyles(MenuItem)
        onClick={[Function]}
      >
        ...
      </WithStyles(MenuItem)>
    </WithStyles(Menu)>
  </div>
</div>
`;

exports[`CommunicatorDialogButtons tests menu behavior 2`] = `
<div
  className="CommunicatorDialogButtons__container"
>
  <div
    className="CommunicatorDialogButtons__searchButton"
  >
    <IconButton
      id="communicator-dialog-buttons-search-button"
      label="cboard.components.CommunicatorDialog.search"
      onClick={[Function]}
    >
      <pure(Search) />
    </IconButton>
  </div>
  <div
    className="CommunicatorDialogButtons__menu"
  >
    <IconButton
      id="communicator-dialog-buttons-menu-button"
      label="cboard.components.CommunicatorDialog.menu"
      onClick={[Function]}
    >
      <pure(MoreVert) />
    </IconButton>
    <WithStyles(Menu)
      anchorEl="someElement"
      id="communicator-dialog-buttons-menu"
      onClose={[Function]}
      open={true}
    >
      <WithStyles(MenuItem)
        onClick={[Function]}
      >
        ...
      </WithStyles(MenuItem)>
    </WithStyles(Menu)>
  </div>
</div>
`;

exports[`CommunicatorDialogButtons tests search behavior 1`] = `
<div
  className="CommunicatorDialogButtons__container"
>
  <div
    className="CommunicatorDialogButtons__searchButton"
  >
    <IconButton
      id="communicator-dialog-buttons-search-button"
      label="cboard.components.CommunicatorDialog.search"
      onClick={[Function]}
    >
      <pure(Search) />
    </IconButton>
  </div>
  <div
    className="CommunicatorDialogButtons__menu"
  >
    <IconButton
      id="communicator-dialog-buttons-menu-button"
      label="cboard.components.CommunicatorDialog.menu"
      onClick={[Function]}
    >
      <pure(MoreVert) />
    </IconButton>
    <WithStyles(Menu)
      anchorEl={null}
      id="communicator-dialog-buttons-menu"
      onClose={[Function]}
      open={false}
    >
      <WithStyles(MenuItem)
        onClick={[Function]}
      >
        ...
      </WithStyles(MenuItem)>
    </WithStyles(Menu)>
  </div>
</div>
`;
